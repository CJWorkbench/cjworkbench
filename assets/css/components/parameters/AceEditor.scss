.code-editor {
  .ace-aspect-ratio-container {
    margin-top: 1.6rem;
    position: relative;
    width: 100%;
    height: 0;
    padding-bottom: 100%; // square

    &>* {
      position: absolute;
      top: 0;
      left: 0;
      width: 100%;
      height: 100%;
    }
  }
}
