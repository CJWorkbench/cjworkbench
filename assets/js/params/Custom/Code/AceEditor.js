import React from 'react'
import PropTypes from 'prop-types'
import AceEditor from 'react-ace/lib/ace'
import memoize from 'memoize-one'
import { Trans } from '@lingui/macro'

import 'brace/mode/python'
import 'brace/theme/xcode'

// Globals -- so each render(), they're equal according to ===
const EditorProps = {
  // $blockScrolling fixes a console.warn() we'd otherwise see
  $blockScrolling: Infinity
}

export default class WorkbenchAceEditor extends React.PureComponent {
  static propTypes = {
    // When isZenMode changes, we'll call componentDidUpdate()
    isZenMode: PropTypes.bool.isRequired,
    wfModuleError: PropTypes.string, // hopefully null or empty
    name: PropTypes.string.isRequired,
    value: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired // func(value) => undefined
  }

  state = {
    // We'll modify width and height to px values, causing a change, so
    // AceEditor will know to recalculate things (like line wraps).
    width: '100%',
    height: '100%'
  }

  wrapperRef = React.createRef()

  componentDidMount () {
    this.updateSize()
  }

  componentDidUpdate (prevProps) {
    // ignore state changes, since we _cause_ them
    if (prevProps === this.props) return

    this.updateSize()
  }

  updateSize () {
    const div = this.wrapperRef.current
    if (!div) return

    this.setState({
      width: div.clientWidth + 'px',
      height: div.clientHeight + 'px'
    })
  }

  getAnnotations = memoize(wfModuleError => {
    const m = /^Line (\d+): (.*)/.exec(wfModuleError)
    if (!m) {
      return []
    } else {
      return [
        {
          row: +m[1] - 1,
          type: 'error',
          text: m[2]
        }
      ]
    }
  })

  // Render editor
  render () {
    const { value, onChange, isZenMode, wfModuleError } = this.props

    return (
      <>
        <div className='help'>
          <Trans id='js.params.Custom.Code.AceEditor.help' description='The tags <3>, <6>, and <8> are URLs. The rest are code formatting. Please keep code and names of libraries untranslated.'>
              Define a <kbd>process(table)</kbd> function that accepts
              a <kbd>pd.DataFrame</kbd> and returns
              a <kbd>pd.DataFrame</kbd>. You may use
              the <a target='_blank' rel='noopener noreferrer' href='https://docs.python.org/3/library/math.html'><kbd>math</kbd></a>, <kbd>pd</kbd>{' '}
              (<a target='_blank' rel='noopener noreferrer' href='https://pandas.pydata.org/pandas-docs/stable/api.html#dataframe'>Pandas</a>) {' '}
              and <kbd>np</kbd>{' '}
              (<a target='_blank' rel='noopener noreferrer' href='https://docs.scipy.org/doc/numpy/reference/routines.html'>Numpy</a>) modules.
          </Trans>
        </div>
        <div className='ace-aspect-ratio-container'>
          <div className='ace-wrapper' ref={this.wrapperRef}>
            <AceEditor
              editorProps={EditorProps}
              width={this.state.width}
              height={this.state.height}
              mode='python'
              theme='xcode'
              wrapEnabled
              annotations={this.getAnnotations(wfModuleError)}
              showGutter={isZenMode}
              name='code-editor'
              onChange={onChange}
              value={value}
            />
          </div>
        </div>
      </>
    )
  }
}
