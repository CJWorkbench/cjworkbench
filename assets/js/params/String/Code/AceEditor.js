import { createRef, PureComponent } from 'react'
import PropTypes from 'prop-types'
import AceEditor from 'react-ace/lib/ace'
import memoize from 'memoize-one'
import { Trans } from '@lingui/macro'
import { QuickFixPropTypes } from '../../../WorkflowEditor/step/QuickFix'

import 'ace-builds/src-noconflict/mode-python'
import 'ace-builds/src-noconflict/mode-sql'
import 'ace-builds/src-noconflict/theme-xcode'

// Globals -- so each render(), they're equal according to ===
const EditorProps = {
  // $blockScrolling fixes a console.warn() we'd otherwise see
  $blockScrolling: Infinity
}

function SyntaxHelp ({ syntax }) {
  switch (syntax) {
    case 'python':
      return (
        <Trans
          id='js.params.Custom.Code.AceEditor.help'
          comment='The tags <3>, <6>, and <8> are URLs. The rest are code formatting. Please keep code and names of libraries untranslated.'
        >
          Define a <kbd>process(table)</kbd> function that accepts a{' '}
          <kbd>pd.DataFrame</kbd> and returns a <kbd>pd.DataFrame</kbd>. You may
          use the{' '}
          <a
            target='_blank'
            rel='noopener noreferrer'
            href='https://docs.python.org/3/library/math.html'
          >
            <kbd>math</kbd>
          </a>
          , <kbd>pd</kbd> (
          <a
            target='_blank'
            rel='noopener noreferrer'
            href='https://pandas.pydata.org/pandas-docs/stable/api.html#dataframe'
          >
            Pandas
          </a>
          ) and <kbd>np</kbd> (
          <a
            target='_blank'
            rel='noopener noreferrer'
            href='https://docs.scipy.org/doc/numpy/reference/routines.html'
          >
            Numpy
          </a>
          ) modules.
        </Trans>
      )
    case 'sql':
      return (
        <Trans id='js.params.Custom.Code.AceEditor.help.sql'>
          Write an SQL <kbd>SELECT</kbd> query that reads from the{' '}
          <kbd>"input"</kbd> table.
        </Trans>
      )
  }
}

export default class WorkbenchAceEditor extends PureComponent {
  static propTypes = {
    // When isZenMode changes, we'll call componentDidUpdate()
    isZenMode: PropTypes.bool.isRequired,
    stepOutputErrors: PropTypes.arrayOf(
      PropTypes.shape({
        message: PropTypes.string.isRequired,
        quickFixes: PropTypes.arrayOf(PropTypes.shape(QuickFixPropTypes))
          .isRequired
      }).isRequired
    ).isRequired, // may (hopefully) be empty
    name: PropTypes.string.isRequired,
    value: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired // func(value) => undefined
  }

  state = {
    // We'll modify width and height to px values, causing a change, so
    // AceEditor will know to recalculate things (like line wraps).
    width: '100%',
    height: '100%'
  }

  wrapperRef = createRef()

  componentDidMount () {
    this.updateSize()
  }

  componentDidUpdate (prevProps) {
    // ignore state changes, since we _cause_ them
    if (prevProps === this.props) return

    this.updateSize()
  }

  updateSize () {
    const div = this.wrapperRef.current
    if (!div) return

    // Force height to equal width
    this.setState({
      width: div.clientWidth + 'px',
      height: div.clientWidth + 'px'
    })
  }

  getAnnotations = memoize(stepOutputErrors => {
    return stepOutputErrors
      .map(error => {
        const m = /^Line (\d+): (.*)/.exec(error.message)
        if (m) {
          return {
            row: +m[1] - 1,
            type: 'error',
            text: m[2]
          }
        } else {
          return null
        }
      })
      .filter(x => x)
  })

  // Render editor
  render () {
    const {
      value,
      onChange,
      isZenMode,
      stepOutputErrors,
      placeholder,
      syntax
    } = this.props

    return (
      <div className='ace-wrapper-outer'>
        <div className='help'>
          <SyntaxHelp syntax={syntax} />
        </div>
        <div className='ace-aspect-ratio-container'>
          <div className='ace-wrapper' ref={this.wrapperRef}>
            <AceEditor
              editorProps={EditorProps}
              width={this.state.width}
              height={this.state.height}
              mode={syntax}
              placeholder={placeholder}
              theme='xcode'
              wrapEnabled
              annotations={this.getAnnotations(stepOutputErrors)}
              showGutter={isZenMode /* false hides annotations */}
              name='code-editor'
              onChange={onChange}
              value={value}
            />
          </div>
        </div>
      </div>
    )
  }
}
