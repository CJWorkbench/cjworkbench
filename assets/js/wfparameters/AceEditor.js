import React from 'react'
import PropTypes from 'prop-types'
import AceEditor from 'react-ace/lib/ace'
import memoize from 'memoize-one'

import 'brace/mode/python'
import 'brace/theme/tomorrow'

// Globals -- so each render(), they're equal according to ===
const EditorProps = {
  // $blockScrolling fixes a console.warn() we'd otherwise see
  $blockScrolling: Infinity
}

export default class WorkbenchAceEditor extends React.PureComponent {
  static propTypes = {
    // When isZenMode changes, we'll call componentDidUpdate()
    isZenMode: PropTypes.bool.isRequired,
    wfModuleError: PropTypes.string,
    name: PropTypes.string.isRequired,
    defaultValue: PropTypes.string.isRequired,
    save: PropTypes.func.isRequired // func(value) => undefined
  }

  constructor (props) {
    super(props)

    this.state = {
      value: this.props.defaultValue,
      // We'll modify width and height to px values, causing a change, so
      // AceEditor will know to recalculate things (like line wraps).
      width: '100%',
      height: '100%'
    }

    this.wrapperRef = React.createRef()
  }

  componentDidMount() {
    this.updateSize()
  }

  componentDidUpdate(prevProps) {
    // ignore state changes, since we _cause_ them
    if (prevProps === this.props) return

    this.updateSize()
  }

  updateSize() {
    const div = this.wrapperRef.current
    if (!div) return

    this.setState({
      width: div.clientWidth + 'px',
      height: div.clientHeight + 'px'
    })
  }

  onChange = (newValue) => {
    this.setState({
      value: newValue
    })
  }

  onBlur = () => {
    this.props.save(this.state.value)
  }

  getAnnotations = memoize(wfModuleError => {
    const m = /^Line (\d+): (.*)/.exec(wfModuleError)
    if (!m) {
      return []
    } else {
      return [
        {
          row: +m[1] - 1,
          type: 'error',
          text: m[2]
        }
      ]
    }
  })

  // Render editor
  render () {
    return (
      <div className='code-editor'>
        <div className='label-margin t-d-gray content-3'>{this.props.name}</div>
        <div className='ace-aspect-ratio-container'>
          <div className='ace-wrapper' ref={this.wrapperRef}>
            <AceEditor
              editorProps={EditorProps}
              width={this.state.width}
              height={this.state.height}
              mode='python'
              theme='tomorrow'
              wrapEnabled={true}
              annotations={this.getAnnotations(this.props.wfModuleError)}
              showGutter={this.props.isZenMode}
              name='code-editor'
              onChange={this.onChange}
              onBlur={this.onBlur}
              value={this.state.value}
            />
          </div>
        </div>
      </div>
    )
  }
}
