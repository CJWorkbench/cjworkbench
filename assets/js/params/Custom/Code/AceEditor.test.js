/* global describe, it, expect, jest */
import React from 'react'
import WorkbenchAceEditor from './AceEditor'
import { shallow } from 'enzyme'
import AceEditor from 'react-ace/lib/ace'

describe('AceEditor', () => {
  const defaultProps = {
    name: 'code',
    value: 'def process(table)\n    return table',
    onChange: jest.fn(),
    isZenMode: false
  }

  it('matches snapshot', () => {
    const wrapper = shallow(
      <WorkbenchAceEditor {...defaultProps} save={jest.fn()} />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('annotates an error', () => {
    const wrapper = shallow(
      <WorkbenchAceEditor {...defaultProps} save={jest.fn()} wfModuleOutputError='Line 1: Foo happened' />
    )
    expect(wrapper.find(AceEditor).prop('annotations'))
      .toEqual([{ row: 0, type: 'error', text: 'Foo happened' }])
  })
})
