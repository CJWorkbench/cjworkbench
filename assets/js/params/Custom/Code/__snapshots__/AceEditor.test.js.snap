// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AceEditor matches snapshot 1`] = `
<Fragment>
  <div
    className="help"
  >
    <WithI18n
      components={
        Array [
          <kbd />,
          <kbd />,
          <kbd />,
          <a
            href="https://docs.python.org/3/library/math.html"
            rel="noopener noreferrer"
            target="_blank"
          />,
          <kbd />,
          <kbd />,
          <a
            href="https://pandas.pydata.org/pandas-docs/stable/api.html#dataframe"
            rel="noopener noreferrer"
            target="_blank"
          />,
          <kbd />,
          <a
            href="https://docs.scipy.org/doc/numpy/reference/routines.html"
            rel="noopener noreferrer"
            target="_blank"
          />,
        ]
      }
      defaults="Define a <0>{0}</0> function that accepts a <1>{1}</1> and returns a <2>{2}</2>. You may use the <3><4>{3}</4></3>,<5>{4}</5> (<6>{5}</6>) and <7>{6}</7> (<8>{7}</8>) modules."
      description="The parameters contain either code or names of libraries. The tags are either <kbd> formatting or URLs."
      id="js.params.Custom.Code.AceEditor.help"
      values={
        Object {
          "0": "process(table)",
          "1": "pd.DataFrame",
          "2": "pd.DataFrame",
          "3": "math",
          "4": "pd",
          "5": "Pandas",
          "6": "np",
          "7": "Numpy",
        }
      }
    />
  </div>
  <div
    className="ace-aspect-ratio-container"
  >
    <div
      className="ace-wrapper"
    >
      <ReactAce
        annotations={Array []}
        cursorStart={1}
        editorProps={
          Object {
            "$blockScrolling": Infinity,
          }
        }
        enableBasicAutocompletion={false}
        enableLiveAutocompletion={false}
        enableSnippets={false}
        focus={false}
        fontSize={12}
        height="100%"
        highlightActiveLine={true}
        maxLines={null}
        minLines={null}
        mode="python"
        name="code-editor"
        navigateToFileEnd={true}
        onChange={[MockFunction]}
        onLoad={null}
        onPaste={null}
        onScroll={null}
        placeholder={null}
        readOnly={false}
        scrollMargin={
          Array [
            0,
            0,
            0,
            0,
          ]
        }
        setOptions={Object {}}
        showGutter={false}
        showPrintMargin={true}
        style={Object {}}
        tabSize={4}
        theme="xcode"
        value="def process(table)
    return table"
        width="100%"
        wrapEnabled={true}
      />
    </div>
  </div>
</Fragment>
`;
